### PR TITLE
add option to idlpp to maintain the include file namespace

### PR DESCRIPTION
**Pull request by MarcelJordense**
Depending on the version of OpenSplice add the -o maintain-include-namespace option to idlpp to enable preservation of the include path names in the for the idl generated files.

### CODE REVIEW

File: rosidl_typesupport_opensplice_cpp/rosidl_typesupport_opensplice_cpp/__init__.py [hunk 1]
```diff
@@ -24,6 +25,28 @@
 from rosidl_parser import validate_field_types
 
 
+def check_idlpp_supports_include_namespaces(idl_pp):
+    res = 0
```
Comment (dirk-thomas) [inline rosidl_typesupport_opensplice_cpp/rosidl_typesupport_opensplice_cpp/__init__.py]: This should be initialized with `False`.

File: rosidl_typesupport_opensplice_cpp/rosidl_typesupport_opensplice_cpp/__init__.py [hunk 2]
```diff
@@ -24,6 +25,28 @@
...
+        if m and m.lastindex == 2:
+            major = m.group(1)
+            minor = m.group(2)
+            res = (major > '6.9' or (major == '6.9' and minor > '181127'))
+    else:
+        m = re.search('([1-9][0-9]*\.[0-9]+)\.([0-9]*)', ospl_version)
+        if m and m.lastindex == 2:
+            major = m.group(1)
+            minor = m.group(2)
+            res = (major > '6.10' or (major == '6.10' and minor > '1'))
```
Comment (dirk-thomas) [inline rosidl_typesupport_opensplice_cpp/rosidl_typesupport_opensplice_cpp/__init__.py]: It looks like that the majority of the two blocks are identical. Please move the generic part above the condition and only make the version number comparison line conditional.

**Review by dirk-thomas — COMMENTED** [review]
Two minor comments on the patch itself.

As mentioned in https://github.com/ros2/rmw_opensplice/issues/259#issuecomment-467577374 the major problem is that idlpp segfaults with that new option for me.

**Comment by MarcelJordense** [comment]
@dirk-thomas I have update the code with your comments

File: rosidl_typesupport_opensplice_cpp/rosidl_typesupport_opensplice_cpp/__init__.py [hunk 3]
```diff
@@ -24,6 +25,20 @@
...
+def check_idlpp_supports_include_namespaces(idl_pp):
+    r = subprocess.getstatusoutput(idl_pp + ' -v')
+    ospl_version = r[1].split(':')[1].strip()
+    m = re.search('([1-9][0-9]*\.[0-9]+)\.([0-9]*)', ospl_version)
+    if m and m.lastindex == 2:
+        major = m.group(1)
+        minor = m.group(2)
+
+    if ospl_version.endswith('OSS'):
+        return major > '6.9' or (major == '6.9' and minor > '181127')
```
Comment (dirk-thomas) [inline rosidl_typesupport_opensplice_cpp/rosidl_typesupport_opensplice_cpp/__init__.py]: This comparisons of the `major` variable are incorrect: e.g. `major > '6.9'`

For `major` being `6.10` the expression evaluates to `False` where it should be `True.

Comment (dirk-thomas) [reply rosidl_typesupport_opensplice_cpp/rosidl_typesupport_opensplice_cpp/__init__.py]: As an alternative we could introduce a version dependency to require the latest release when backporting this patch to Crystal? That would avoid the need for the version check and always use `-o maintain-include-namespace`. We could also remove the extra include directories in that case.

Comment (dirk-thomas) [reply rosidl_typesupport_opensplice_cpp/rosidl_typesupport_opensplice_cpp/__init__.py]: @MarcelJordense Please see #21 for the described alternative. What do you think?

**Review by dirk-thomas — APPROVED** [review]

**Comment by dirk-thomas** [comment]
On Linux still backward compatible with the old OpenSplice release:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/6295/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2713/)

On macOS and Windows the previous release of OpenSplice is still being used:
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5173/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6132/)

**Comment by dirk-thomas** [comment]
Reverted in #22.

**Pull request merged**